### PR TITLE
tammy/move_vcftools_to_ugbio

### PR DESCRIPTION
**Pull request by TammushB**

### CODE REVIEW

File: src/core/tests/unit/vcfbed/test_vcftools.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            "x_gcc",
+            "x_ic",
+            "x_il",
+            "x_lm",
+            "x_rm",
+        ]
+
+    def test_get_vcf_df_use_qual(self, resources_dir):
+        input_vcf = pjoin(resources_dir, "test_get_vcf_df.vcf.gz")
+        df = vcftools.get_vcf_df(input_vcf, scoring_field="QUAL")  # noqa: PD901
```
Comment (avigailmo) [inline src/core/tests/unit/vcfbed/test_vcftools.py:160]: why not to fix this [PD901](https://docs.astral.sh/ruff/rules/pandas-df-variable-name/) error? here and in other places

File: src/core/tests/unit/vcfbed/test_vcftools.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    assert all(hmer_fn["indel"])
+    assert all(x == 1 for x in hmer_fn["hmer_indel_length"])
+    hmer_fn = vcftools.FilterWrapper(data).get_h_mer(val_start=3, val_end=10).get_df()
+    assert all(hmer_fn["indel"])
+    assert all(x >= 3 and x <= 10 for x in hmer_fn["hmer_indel_length"])
+
+    # hmer_fp testing
+    hmer_fp = vcftools.FilterWrapper(data).get_h_mer().get_fp().get_df()
+    assert all(hmer_fp["indel"])
+    assert all([x > 0 for x in hmer_fp["hmer_indel_length"]])  # noqa: C419
```
Comment (avigailmo) [inline src/core/tests/unit/vcfbed/test_vcftools.py:47]: same [C419](https://docs.astral.sh/ruff/rules/unnecessary-comprehension-in-call/) should be easy to fix

File: src/core/ugbio_core/vcfbed/vcftools.py [hunk 3]
```diff
@@ -0,0 +1,772 @@
...
+    info_format : tuple
+        Tuple of info format (Tag, # values, format, Description)
+    """
+    with pysam.VariantFile(vcf_input_file) as vcfin:
+        hdr = vcfin.header
+        if info_format[0] not in hdr.info.keys():
+            hdr.info.add(*info_format)
+        with pysam.VariantFile(vcf_output_file, mode="w", header=hdr) as vcfout:
+            for row in vcfin:
+                val = df.loc[[(row.chrom, row.start + 1)]][column].values[0]  # noqa: PD011
```
Comment (avigailmo) [inline src/core/ugbio_core/vcfbed/vcftools.py:245]: same: https://docs.astral.sh/ruff/rules/pandas-use-of-dot-values/

File: src/core/ugbio_core/vcfbed/vcftools.py [hunk 4]
```diff
@@ -0,0 +1,772 @@
...
+        custom_info_fields = []
+
+    header = pysam.VariantFile(variant_calls).header
+    if chromosome is None:
+        variant_file = pysam.VariantFile(variant_calls)
+
+    else:
+        variant_file = pysam.VariantFile(variant_calls).fetch(chromosome)
+    # pylint: disable-next=bad-builtin
+    vfi = map(  # noqa: C417
```
Comment (avigailmo) [inline src/core/ugbio_core/vcfbed/vcftools.py:66]: same: https://docs.astral.sh/ruff/rules/unnecessary-map/

File: src/core/ugbio_core/vcfbed/vcftools.py [hunk 5]
```diff
@@ -0,0 +1,772 @@
...
+    """
+    initial_region = (max(vpos - region_size // 2, 0), vpos + region_size // 2)
+    if len(vlocs) == 0:
+        return initial_region
+
+    # expand the region to the left
+    # clip for the cases when the variant is after all the variants and need
+    # to be inserted at len(vlocs)
+    while (
+        vlocs[np.clip(np.searchsorted(vlocs, initial_region[0]), 0, len(vlocs)) - 1] - initial_region[0] < 10  # noqa: PLR2004
```
Comment (avigailmo) [inline src/core/ugbio_core/vcfbed/vcftools.py:282]: https://docs.astral.sh/ruff/rules/magic-value-comparison/

File: src/core/ugbio_core/vcfbed/vcftools.py [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+    ):
+        initial_region = (initial_region[0] - 10, initial_region[1])
+
+    initial_region = (max(initial_region[0], 0), initial_region[1])
+
+    # expand the region to the right
+    # The second conditions is for the case np.searchsorted(vlocs,
+    # initial_region[1]) == 0
+    while (
+        initial_region[1] - vlocs[np.clip(np.searchsorted(vlocs, initial_region[1]), 1, len(vlocs)) - 1] < 10  # noqa: PLR2004
```
Comment (avigailmo) [inline src/core/ugbio_core/vcfbed/vcftools.py:293]: https://docs.astral.sh/ruff/rules/magic-value-comparison/

File: src/core/ugbio_core/vcfbed/vcftools.py [hunk 7]
```diff
@@ -0,0 +1,772 @@
...
+                "thickStart",
+                "thickEnd",
+                "itemRgb",
+            ]
+        self.df = self.df.sort_values(["chrom", "chromStart"])
+
+        return self
+
+
+def bed_files_output(data: pd.DataFrame, output_file: str, mode: str = "w", create_gt_diff: bool = True) -> None:  # noqa: FBT001, FBT002
```
Comment (avigailmo) [inline src/core/ugbio_core/vcfbed/vcftools.py:502]: fix precommit errors

File: src/core/tests/unit/vcfbed/test_vcftools.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        vcounts = []
+        bcftools_out = subprocess.check_output(f"bcftools index -s {output_vcf}", shell=True)
+        bcftools_lines = bcftools_out.decode().strip().split("\n")
+        for line in bcftools_lines:
+            columns = line.split("\t")
+            vcounts.append(int(columns[2]))
+        assert vcounts == [653, 227, 523, 1147]
+
+
+# TODO: tests for subsample_to_alleles and header_record_number
```
[the remark this answers is not in the document]
Comment (TammushB) [reply src/core/tests/unit/vcfbed/test_vcftools.py:262]: taken from VC. leaving it as is.

**Review by avigailmo — APPROVED** [review]
great work
few more noqas can still be fixed I think
check it in your other pr

**Comment by TammushB** [comment]
> great work few more noqas can still be fixed I think check it in your other pr

Done

**Pull request merged**